File: 2019/1/puzzle.js
const fs = require('fs');
const fileName = 'input.txt';

function calculateFuel(mass) {
  return Math.floor(mass/3) - 2;
}


function advancedCalculateFuel(mass) {
  total = 0;
  delta = calculateFuel(mass);
  while (delta >= 0) {
    total += delta;
    delta = calculateFuel(delta);
  }
  return total;
}


fs.readFile(fileName, 'utf8', function(err, data) {
  if (err) throw err;
  const masses = data.split('\n').map(x => parseInt(x));
  masses.pop();
  let part1 = 0;


  for (let i = 0; i < masses.length; i++) {
    part1 += calculateFuel(masses[i])
  }
  console.log(part1);

  let part2 = 0;
  for (let i = 0; i < masses.length; i++) {
    part2 += advancedCalculateFuel(masses[i]);
  }
  console.log(part2);
});

File: 2019/1/puzzle.py
with open('input.txt') as f:
    masses = f.readlines()

masses = [int(x) for x in masses]

test_input = (12, 14, 1969, 100756)
test_output = (2, 2, 654, 33583)


def calculate_fuel(mass):
    return mass // 3 - 2

print(sum([calculate_fuel(x) for x in masses]))


def adv_calculate_fuel(mass):
    total = 0
    delta = calculate_fuel(mass)
    while delta >= 0:
        total += delta
        delta = calculate_fuel(delta)
    return total

print(sum([adv_calculate_fuel(x) for x in masses]))
